torch/_dynamo/compile_package.py

```diff
@@ -0,0 +1,756 @@
+"""
+This module provides the infrastructure for creating and managing compile package
+for torch.compile. We mainly have two abstractions here:
+  - Precompile: Compiled artifacts related to one particular recompile.
+  - CompilePackage: Overarching data structure for store and lookup a list of precompiles.
+
+This is different from the typical global caching system in the sense that compile package is
+always saved/loaded via user API calls. This means the caching behavior is always under
+user control explicitly so that a stronger guarantee can be provided about cache hit for a
+specific compiled model. Users can load the compile package from a different process or even
+host but cautions should be taken that compile package will only check a subset of the original
+Dynamo guards so there might be soundness problems.
+"""
+
+import contextlib
+import dataclasses
+import functools
+import glob
+import importlib
+import io
+import logging
+import os
+import pickle
+import platform
+import shutil
+import types
+from collections.abc import Generator
+from dataclasses import dataclass
+from typing import Any, Callable, Optional, TYPE_CHECKING
+
+import torch
+import torch._inductor.package
+from torch._guards import Source
+
+from .bytecode_transformation import get_code_keys
+
+
+logger = logging.getLogger(__name__)
+
+if TYPE_CHECKING:
+    from torch._functorch._aot_autograd.autograd_cache import AOTAutogradCacheEntry
+    from torch._inductor.output_code import CompiledAOTI
+
+
+@dataclasses.dataclass(frozen=True)
+class SerializedCode:
+    co_argcount: int
+    co_posonlyargcount: int
+    co_kwonlyargcount: int
+    co_nlocals: int
+    co_stacksize: int
+    co_flags: int
+    co_code: bytes
+    co_consts: tuple[Any, ...]
+    co_names: tuple[str, ...]
+    co_varnames: tuple[str, ...]
+    co_filename: str
+    co_name: str
+    co_firstlineno: int
+    co_linetable: bytes
+    co_cellvars: tuple[str, ...]
+    co_freevars: tuple[str, ...]
+    co_qualname: Optional[str] = None
+    co_exceptiontable: Optional[bytes] = None
+
+
+class _GraphCompile:
+    """
+    Stores the compiled artifacts per compiled FX graph. This includes:
+      - AOTI compiled kernels.
+      - AOTAuograd wrappers.
+    """
+
+    def __init__(self, name: str) -> None:
+        self.name = name
+        self.forward_aoti: Optional[CompiledAOTI] = None
```

Comment:
In my head, what would make most sense is that an AOTAutogradCacheEntry contains a forward/backward_aoti. I know this is a prototype so you probably didn't want to change AOTAutogradCacheEntry, but I think what you'd actually want is:
- Refactor AOTAutogradCacheEntry into an interface witha  forward/backward OutputCode + the aot autograd wrappers. You could make it generic on the OutputCode type, i.e. AOTAutogradCacheEntry[CompiledAOTI] vs. AOTAutogradCacheEntry[CompiledFXGraph], etc. 
 
- Split AOTAutogradCacheEntry into the cache path, where there's a python wrapper + FXGraphCache keys, vs. the package API, which uses AOTI artifacts

Then here, you would just store self.aot_autograd_output : AOTAutogradCacheEntry[CompiledAOTI].